### ISSUE TITLE
Elixir 1.8 EEx "regression"

### ISSUE THREAD

**Issue by axelson** [elixir-lang/elixir#8790]
In Elixir 1.8 the internals of the EEx engine have been changed in a way that was meant to be backwards compatible (since downstream projects shouldn't be relying on the internals). I am highlighting this issue in the hope that there is some way to improve the error message in Elixir itself. I've written a longer write-up in the phoenix_html project: https://github.com/phoenixframework/phoenix_html/issues/247

### Environment

* Elixir & Erlang/OTP versions (elixir --version):
```
Erlang/OTP 21 [erts-10.2.3] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]

Elixir 1.8.1 (compiled with Erlang/OTP 21)
```
* Operating system: Arch Linux (but affects others as well)

### Current behavior

Running with phoenix_html 2.10.2

Code:
```
<section class="phx-hero"></section>

<%= form_tag("/", method: "post") do %>
  <div class='control-group'></div>
<% end %>
```

Error:
```
== Compilation error in file lib/my_example_web/views/page_view.ex ==
** (BadMapError) expected a map, got: {:safe, [{:|, [], ["", "<section class=\"phx-hero\"></section>\n\n"]}]}
    (eex) lib/eex/engine.ex:161: EEx.Engine.handle_begin/1
    (eex) lib/eex/compiler.ex:54: EEx.Compiler.generate_buffer/4
    (phoenix) lib/phoenix/template.ex:349: Phoenix.Template.compile/3
    (phoenix) lib/phoenix/template.ex:160: anonymous fn/4 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1940: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    lib/my_example_web/views/page_view.ex:1: MyExampleWeb.PageView (module)
    (elixir) lib/kernel/parallel_compiler.ex:208: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
```

### Expected behavior

Ideally the EEx template should render cleanly or a better error message should be provided. Although it is likely that neither of those are feasible in which case this issue should simply be closed.

**Comment by josevalim** [comment]
Hi @axelson! Thank you so much for looking into this. I think we can do a quick check on handle_begin/handle_end/handle_body that the state is actually the one defined on init and raise a meaningful error message in case it isn't.

**Comment by axelson** [comment]
@josevalim thanks! That sounds like a good approach.

**Comment by josevalim** [comment]
@axelson btw, we would love if you could send a PR. If you can't, please let us know and we will tackle it eventually. Thanks!

**Comment by axelson** [comment]
Noting that I do plan to create a PR for this

**Comment by josevalim** [comment]
Closing in favor of #8976. A new v1.8 release will be out once it is merged.
